### PR TITLE
Fix condition to use slice copy in ndarray.__setitem__

### LINKED ISSUE #3013: [Bug] Mask assignment throws an error when CUPY_EXPERIMENTAL_SLICE_COPY=1
An error occurs when setting CUPY_EXPERIMENTAL_SLICE_COPY = 1

  - CuPy version: 6.6.0
  - OS/Platform: Ubuntu 16.04.6 LTS
  - CUDA version: 10.0

* Code to reproduce
```
import os
os.environ["CUPY_EXPERIMENTAL_SLICE_COPY"] = "1"
import cupy as cp

a = cp.array([1,2,3,4,5])
a[a<3] = 0
a
```
* Error messages, stack traces, or logs
```
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-1-9ff589a008e4> in <module>
      4 
      5 a = cp.array([1,2,3,4,5])
----> 6 a[a<3] = 0
      7 a

cupy/core/core.pyx in cupy.core.core.ndarray.__setitem__()

cupy/core/core.pyx in cupy.core.core.ndarray.__richcmp__()

cupy/core/_kernel.pyx in cupy.core._kernel.ufunc.__call__()

cupy/core/_kernel.pyx in cupy.core._kernel._preprocess_args()

TypeError: Unsupported type <class 'slice'>
```

### PR DESCRIPTION
**Pull request by takagi**
Fix #3013 .

This PR fixes the condition to use CUPY_EXPERIMENTAL_SLICE_COPY feature in `ndarray.__setitem__` to check if the supplied `slices` is a slice object.

### CODE REVIEW

**Comment by takagi** [comment]
Thanks, I've moved the test as you suggested!

**Comment by takagi** [comment]
Fixed!

File: cupy/core/core.pyx [hunk 1]
```diff
@@ -1233,11 +1233,13 @@ cdef class ndarray:
...
-            type(slices) is slice and slices == slice(None, None, None) and
-            isinstance(value, numpy.ndarray)):
-            if (self.dtype == value.dtype and
-                    self.shape == value.shape):
+        def does_slice_copy(slices, value):
+            return (type(slices) is slice and
+                    slices == slice(None, None, None) and
+                    isinstance(value, numpy.ndarray))
+
+        if util.ENABLE_SLICE_COPY and does_slice_copy(slices, value):
```
Comment (toslunar) [inline cupy/core/core.pyx]: - I'm not sure the overhead of the function call is negligible.
- (nit) Break lines before operators.
```suggestion
        if util.ENABLE_SLICE_COPY and (
                type(slices) is slice
                and slices == slice(None, None, None)
                and isinstance(value, numpy.ndarray)
        ):
```

Comment (takagi) [reply cupy/core/core.pyx]: Sure!

**Comment by toslunar** [comment]
Jenkins, test this please.

**Pull request merged**